File: tools/convert_blog_post/src/anytype_simplied/collection.rs
use std::collections::BTreeMap;

use anyhow::{Error, Ok};
use serde::{Deserialize, Serialize};

use crate::{
    anytype::object::AnytypeObject,
    anytype_proto::{
        anytype::SnapshotWithType,
        trait_impl::{get_field_value, get_field_value_list, to_val_map},
    },
};

use super::{
    attributes::AttributeMap,
    page::{Page, PageId, PageMap},
    workspace::WorkspaceId,
};

// use super::article::Article;

pub type CollectionId = String;
pub type CollectionMap = BTreeMap<CollectionId, Box<Collection>>;

#[derive(Default, Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct Collection {
    pub id: CollectionId,
    pub name: String,
    pub description: String,
    pub cover: String,
    pub styles: Vec<String>,
    #[serde(skip_serializing)]
    pub attributes: AttributeMap,
    #[serde(skip_serializing)]
    pub workspace_id: WorkspaceId,

    pub articale_id_list: Vec<PageId>,
    pub articles: PageMap,
}

impl Collection {
    #[deprecated]
    pub fn from_anytype(raw: &AnytypeObject) -> Self {
        let data = raw.snapshot.data.details.clone();

        let mut articale_id_list: Vec<String> = vec![];
        if let Some(collections) = raw.snapshot.data.collections.clone() {
            articale_id_list = collections.objects.unwrap_or_default();
        }
        Self {
            id: data.id.clone(),
            name: data.name.unwrap_or_default().clone(),
            description: data.description.unwrap_or_default().clone(),
            // cover: data.cover.unwrap_or_default().clone(),
            // styles: data.styles.unwrap_or_default().clone(),
            // articale_id_list: raw.snapshot.data.collections.unwrap_or_default().objects
            articale_id_list,
            ..Default::default()
        }
    }

    pub fn from_anytype_proto(raw: &SnapshotWithType) -> Result<Self, Error> {
        let mut tmp = Self::default();
        // let data = raw.snapshot.data.details.clone();
        let ref data_map = raw
            .snapshot
            .as_ref()
            .unwrap()
            .data
            .as_ref()
            .unwrap()
            .details
            .as_ref()
            .unwrap();

        tmp.id = get_field_value(data_map, "id")?;
        tmp.name = get_field_value(data_map, "name")?;
        tmp.attributes = to_val_map(data_map)?;
        tmp.articale_id_list = get_field_value_list(data_map, "links")?;
        tmp.workspace_id = get_field_value(data_map, "spaceId")?;
        // tmp.description = get_field_value(data_map, "description")?;
        // tmp.serie = get_field_value(data_map, "Serie")?;

        // let mut articale_id_list: Vec<String> = vec![];
        // if let Some(collections) = raw.snapshot.data.collections.clone() {
        //     articale_id_list = collections.objects.unwrap_or_default();
        // }
        Ok(tmp)
    }

    pub fn insert_articles(&mut self, files: &Vec<Page>) -> Result<(), Error> {
        for key in self.articale_id_list.iter() {
            if let Some(file) = files.iter().find(|f| &f.id == key) {
                self.articles.insert(file.id.to_owned(), file.to_owned());
            }
        }
        Ok(())
    }
}

File: tools/convert_blog_post_ts/collection.ts
// import { BTreeMap } from "immutable"; // Assuming you are using Immutable.js for BTreeMap
// import { AnytypeObject, SnapshotWithType, Error } from "./types"; // Adjust the import according to your types
import { castToAttributeMap, type AttributeMap } from "./attribute";
import type { SnapshotWithType } from "../../protos/anytype/snapshot";
import type { WorkspaceId } from "./workspace";
import type { Page, PageMap } from "./page";
import { getFieldValue } from "./common";
import { ObjectTypes } from "./enum_token";
export type CollectionId = string;
export type CollectionMap = Map<CollectionId, Collection>;

// object interface
export interface Collection {
  id: CollectionId;
  name: string;
  description: string;
  cover: string;
  styles: string[];
  attributes: AttributeMap;
  workspace_id: WorkspaceId;
  article_id_list: string[];
  articles: PageMap;

  // fromAnytype?(raw: SnapshotWithType): Collection;
}

export function fromAnytype(raw: SnapshotWithType): Collection {
  const dataMap = raw.snapshot?.data?.details;

  const collection: Collection = {
    id: getFieldValue<string>(dataMap, "id"),
    name: getFieldValue<string>(dataMap, "name"),
    // attributes: getFieldValue<AttributeMap>(dataMap, "attributes"),
    attributes: castToAttributeMap(dataMap),
    article_id_list: getFieldValue<string[]>(dataMap, "links"),
    workspace_id: getFieldValue(dataMap, "spaceId"),
    description: "", // Assuming default value
    cover: "", // Assuming default value
    styles: [], // Assuming default value
    articles: {} as PageMap,
  };

  return collection;
}

export function insertArticles(collection: Collection, files: Page[]): void {
  for (const key of collection.article_id_list) {
    const file = files.find((f) => f.id === key);
    if (file) {
      collection.articles[file.id] = file; // Assuming articles is an object
    }
  }
}

export const getCollections = (
  snapshot_list: SnapshotWithType[],
): Collection[] => {
  return snapshot_list
    .filter((snapshot) =>
      snapshot.snapshot?.data?.objectTypes.includes(
        ObjectTypes.Collection.toString(),
      ),
    )
    .map(fromAnytype);
};
